### DISCUSSION TITLE
Recursive data structure validation

### DISCUSSION THREAD

**Discussion by deanmcpherson** [spatie/laravel-data#73] [Q&A]
I'm trying to use a data structure to validate a recursive document data structure (prosemirror). I'm running into an issue where I need a class to be able to reference a nested collection of itself;

e.g.

```php

class PmMark extends Data {
    public string $type;
}

class PmNode extends Data {
    public string $text;
    public string $type;
     /** @var PmMark[] */
    public ?DataCollection $marks;
    
    /** @var PmNode[] */
     public ?DataCollection $content;

     public ?DataCollection $attrs;

}
class PmDocData extends Data {
    public function __construct(
       #[In('doc')]
       public string $type,
       
        /** @var PmNode[] */
       public DataCollection $content
    )
    {
    }
}
```

When I validate data against this type, it just locks up. I assume it's because the validation would be pre-built, and it's just getting in an infinite loop. Are there any known workarounds?

**Comment by rubenvanassche** [comment]
In such cases it is recommended to manually write validation rules within a `rules` method of a data object, when the rule exists for such a property we won't try to automatically deduce the validation properties which causes the infinite loop.
